ml-agents-envs/mlagents_envs/registry/binary_utils.py

```diff
@@ -0,0 +1,214 @@
+import urllib.request
+import tempfile
+import os
+import uuid
+import shutil
+import glob
+import yaml
+import hashlib
+
+from zipfile import ZipFile
+from sys import platform
+from typing import Tuple, Optional, Dict, Any
+
+from mlagents_envs.logging_util import get_logger
+
+logger = get_logger(__name__)
+
+# The default logical block size is 8192 bytes (8 KB) for UFS file systems.
+BLOCK_SIZE = 8192
+
+
+def get_local_binary_path(name: str, url: str) -> str:
+    """
+    Returns the path to the executable previously downloaded with the name argument. If
+    None is found, the executable at the url argument will be downloaded and stored
+    under name for future uses.
+    :param name: The name that will be given to the folder containing the extracted data
+    :param url: The URL of the zip file
+    """
+    path = get_local_binary_path_if_exists(name, url)
+    if path is None:
+        logger.debug(
+            f"Local environment {name} not found, downloading environment from {url}"
+        )
+        download_and_extract_zip(url, name)
```

Comment:
Connection errors will raise from here and aren't caught. Since these might be transient errors and not the users fault, we should catch them and return null (or even better, retry a few times).